### PR TITLE
Lab 2 (by Andrey Khoroshavin)

### PR DESCRIPTION
**Pull request by Andrew174194**
Pull request for lab 2
by Andrey Khoroshavin (kill team, CS-01 academic group)

PS. `Transfer from Lab1` is simple copy from previous lab, so you can check differences by comparing that and last commits.

### CODE REVIEW

File: app_python/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,6 @@
+FROM python:3.9-alpine
```
Comment (DOCtorActoAntohich) [inline app_python/Dockerfile:1]: this dude ruined everything for me :trollface: 
it has literally nothing so when you install some advanced libraries, it'll probably fail cos it has no gcc
(yours works fine tho )

File: app_python/README.md [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+- UNIX-like or Windows system with the correct time set (the timezone can be different from Europe/Moscow)
+- Access to network or Internet (or only localhost address is available) 
+
+## Build
+No building is required for this source code since Python is an interpreted language.
+
+## Usage
+There are 2 ways for running this application:
+### Docker (preferable)
+1. Download the code to your local machine - to be more specific, `main.py` & `requirements.txt` are required
```
Comment (DOCtorActoAntohich) [inline app_python/README.md:23]: i mean, people usually clone the whole repository :trollface:

**Review by DOCtorActoAntohich — APPROVED** [review]
gud

File: app_python/DOCKER.md [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+    I expose only 1 necessary for HTML requests port - 5000, which it written in Dockerfile.
+
+5. Don't put confidential information inside container
+
+    I have no secrets for this project, however it is important to pass secret words with `--secret` parameter.
+
+6. Prefer COPY over ADD
+
+    Because ADD is more error prone than COPY, it is better to use the last one when possible.
```
Comment (ntdesmond) [inline app_python/DOCKER.md:31]: You swapped them and the last one now is `COPY` xd

File: app_python/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,6 @@
...
+EXPOSE 5000
+COPY main.py requirements.txt ./app/
+RUN cd ./app/ && pip install -r requirements.txt && adduser -D app_user && chown -R root ./ && chmod -R 755 ./
+USER app_user
+CMD python ./app/main.py
```
Comment (ntdesmond) [inline app_python/Dockerfile:6]: `chown` and `chmod` are bullshit btw:

- You write here `USER app_user`, and before it reaches this line, everything is run as root already. Thus, `COPY`-ed files are already owned by `root` (UID 0).

- Default permissions for other users, I guess, are also `r-x`, thus `chmod` here also does not change anything, at least for the user `app_user`, which does not own the files

Comment (ntdesmond) [reply app_python/Dockerfile:6]: And yeah, also the thing I copied from my other reviews:

---

Don't use `flask run` in production! Use WSGI servers for that, e.g. [uwsgi](https://github.com/unbit/uwsgi) or [gunicorn](https://github.com/benoitc/gunicorn).

Even better, if you proxy the app afterwards (I didn't do that either )

File: app_python/README.md [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+2. Download `Dockerfile` and use `docker build -t andrew174194/msc_time_py .` to build container by yourself OR
+    use pre-build container with command `docker pull andrew174194/msc_time_py` ([Link to image on Docker Hub](https://hub.docker.com/r/andrew174194/msc_time_py))
+3. Run container by `docker run -p 80:5000 andrew174194/msc_time_py`, the web-application will be available on `localhost` or `localhost:80`. You can change 80 port to other port you like.
```
Comment (ntdesmond) [inline app_python/README.md:26]: If I wanna pull and run the image, I won't need to download the source code. Instructions unclear!

Comment (ntdesmond) [reply app_python/README.md:26]: P.S. I would also split this `OR` to two separate list entries (maybe add a nested list)

**Review by ntdesmond — APPROVED** [review]

File: app_python/README.md (hunk 2 above)
Comment (Andrew174194) [reply app_python/README.md:23]: We have so much "info" stuff to keep in Docker, you know 🙃

**Pull request closed without merging**